### PR TITLE
Implement dark overlay for menus and automap

### PR DESCRIPTION
**Pull request by andrikpowell**
This PR adds the dark overlay that Woof has.

I've implemented it via `V_DrawShaded` which gets directed to two different functions depending on OpenGL or Software (this is how most render functions currently work in DSDA Doom)

The `V_DrawShaded` function includes a gradual fade option (`dboolean animate`) that's used in Nyan Doom, although I've set it to false in both instances that it's used here to match Woof's implementation.

The automap now has 3 overlay settings in the same way Woof does: "No Overlay", "Overlay 1" (what DSDA already had), and "Overlay dark" which adds a translucent background. The dark overlay is also compatible with the DSDA Doom Minimap.

~The final commit here is probably the most controversial, but I wanted to bring it up here. It's because of this commit that implementing the dark overlay was actually harder to get working for DSDA Doom compared to Nyan Doom. In order for Nyan Doom to support overlapping HUD icons and statusbar animations, I have the statusbar always refresh in Software just as it does in OpenGL. In my testing I haven't run into issues or slowdown by doing this, and it fixes many statusbar bleed issues with PWADs such as "Pirate Doom II" and "200 Line Christmas".~

### CODE REVIEW

**Comment by andrikpowell** [comment]
I've done a force-push as I found an issue regarding the overlay menu and STBAR... However this force push fixes all the STBAR issues for Doom and Raven.

**Comment by andrikpowell** [comment]
> The final commit here is probably the most controversial, but I wanted to bring it up here. It's because of this commit that implementing the dark overlay was actually harder to get working for DSDA Doom compared to Nyan Doom. In order for Nyan Doom to support overlapping HUD icons and statusbar animations, I have the statusbar always refresh in Software just as it does in OpenGL. In my testing I haven't run into issues or slowdown by doing this, and it fixes many statusbar bleed issues with PWADs such as "Pirate Doom II" and "200 Line Christmas".

I've decided to extract this and place it into another pull request... It's not that related to this PR, and is honestly a separate issue altogether.

**Comment by rfomin** [comment]
Set to OpenGL mode, go to the general menu and wait a few minutes - the background will start to flicker. On my system it seems to happen more often with VSync turned off.

**Comment by Pedro-Beirao** [comment]
So the map overlay keybind cycles thru these 3 options. I dont like that since I use overlay when playing, but when I want to check the map name, coords, etc, I switch to no overlay.

This will make me have to cycle thru 1 more option. Its one of those small changes that affect the players in annoying ways. I think it would be best to have an option in the automap menu to decide if the overlay is faded or not.

**Comment by Pedro-Beirao** [comment]
[Bug] Being in the faded menu, but then pressing CMD+Q (Mac keybind to bring up the "Are u sure u want to quit" prompt)and then cancelling: keeps the titlepic faded, and is only fixed after pressing Esc to bring up the menu again

This doesnt happen with the Texture option

**Comment by Pedro-Beirao** [comment]
Can you remove the animate logic altogether? Since its always false + it looks better with no fade

**Comment by andrikpowell** [comment]
I've removed *(or extracted for Nyan Doom)* the overlay fade logic. Point is it's now gone for DSDA Doom.

In addition, I've fixed the Mac "CMD + Q" overlay issue.

I spent a full day trying another approach for the overlay logic, but it wasn't very compatible with disabling itself when the automap overlay is present, so I think the current way is better... I really don't wanna talk about how long I spent on this...

I plan on moving onto the revising the automap logic next :)

**Comment by andrikpowell** [comment]
Ready for automap dark overlay review

File: prboom2/src/am_map.c [hunk 1]
```diff
@@ -616,7 +617,7 @@ void AM_SetPosition(void)
     f_x = f_y = 0;
     f_w = SCREENWIDTH;
 
-    if (automap_overlay)
+    if (automap_overlay>0)
```
Comment (rfomin) [inline prboom2/src/am_map.c]: Sorry to be picky, but I think we should try to use a uniform style for the new code: `if (automap_overlay > 0)` with spaces.

The original PrBoom+ was badly formatted, but all the new code from Kraflab has the same style (files in the dsda folder).

File: prboom2/src/m_menu.c [hunk 2]
```diff
@@ -5823,6 +5828,31 @@ void M_StartControlPanel (void)
...
 }
 
+
+/////////////////////////////////////////////////////////////////////////////
+//
+// Menu Shaded Overlay Stuff
+//
+// This displays a dark overlay under certain screens of the menus
+
+int screenshade = 20;
```
Comment (rfomin) [inline prboom2/src/m_menu.c]: We should use `static const` or macro if it is a constant. Do we need this variable?

File: prboom2/src/m_menu.c [hunk 3]
```diff
@@ -1189,81 +1189,6 @@ void M_Options(int choice)
   M_SetupNextMenu(&OptionsDef);
 }
 
-/////////////////////////////
-//
-// M_QuitDOOM
-//
```
Comment (rfomin) [inline prboom2/src/m_menu.c]: Why did you move this code?

**Review by rfomin — COMMENTED** [review]
It works well from my testing, thank you!

File: prboom2/src/am_map.c (hunk 1 above)
Comment (andrikpowell) [reply prboom2/src/am_map.c]: Easy enough to fix 👍

File: prboom2/src/m_menu.c (hunk 2 above)
Comment (andrikpowell) [reply prboom2/src/m_menu.c]: So in Nyan Doom, I have extra code where this variable is not constant... And so I was trying to keep the code close to the same between the two ports, while deleting extra animation fade code for DSDA Doom.

File: prboom2/src/m_menu.c (hunk 3 above)
Comment (andrikpowell) [reply prboom2/src/m_menu.c]: Beside moving this code, I've added the two lines:
```
setup_active = false;
currentMenu = NULL;
```
This is to avoid the overlay getting applied to the quit message when sending the SDL_Quit response while the overlay is open. The `M_QuitDOOM` code needed to be moved down below the `setup_active` declaration to change the variable status.

Comment (rfomin) [reply prboom2/src/m_menu.c]: It's easier to move the `setup_active` declaration to the beginning of the file. If you move a big chunk of code, we can't see in `diff` what was changed in it.

File: prboom2/src/m_menu.c (hunk 2 above)
Comment (rfomin) [reply prboom2/src/m_menu.c]: Are there plans to merge code from Nyan into DSDA-Doom or vice versa?

Comment (andrikpowell) [reply prboom2/src/m_menu.c]: No, but isn't it better to keep the implementation of the same feature in both source ports as close as possible? As in if I add something to Nyan that DSDA wants in the future or vice-versa, it will be easier if DSDA is already using a very similar structure.

I mention this because there are certain things I can "optimise / simplify", but I haven't because I wanted to keep the structure relatively close to Nyan for consistency sake.

File: prboom2/src/m_menu.c (hunk 3 above)
Comment (andrikpowell) [reply prboom2/src/m_menu.c]: Decided to move back `M_QuitDOOM` and move the menu active declarations to the top of the file. I'd rather not separate them, as the comments already have them grouped together.

File: prboom2/src/am_map.c (hunk 1 above)
Comment (andrikpowell) [reply prboom2/src/am_map.c]: Should be resolved in latest commit

File: prboom2/src/m_menu.c [hunk 4]
```diff
@@ -5823,6 +5828,31 @@ void M_StartControlPanel (void)
...
+
+dboolean M_MenuIsShaded(void)
+{
+  int Options = (setup_active || currentMenu == &OptionsDef || currentMenu == &SoundDef);
+  return fadeBG() && Options;
+}
+
+static void M_ShadedScreen(int scrn)
+{
+  V_DrawShaded(scrn, 0, 0, SCREENWIDTH, SCREENHEIGHT, screenshade);
```
Comment (rfomin) [inline prboom2/src/m_menu.c]: In general, it is good practice to reduce the number of global variables as much as possible. Doom is not good at this, but let's not make it worse. Let's remove the global variable `screenshade`, but keep the function parameter `screenshade`, then the difference with Nyan Doom would be minimal:
```diff
diff --git a/prboom2/src/am_map.c b/prboom2/src/am_map.c
index 3650a72c9..eeda64153 100644
--- a/prboom2/src/am_map.c
+++ b/prboom2/src/am_map.c
@@ -2801,7 +2801,7 @@ void AM_Drawer (dboolean minimap)
   if (!automap_overlay) // cph - If not overlay mode, clear background for the automap
     V_FillRect(FB, f_x, f_y, f_w, f_h, (byte)mapcolor_p->back); //jff 1/5/98 background default color
   if (automap_overlay == 2 && !M_MenuIsShaded())
-    V_DrawShaded(FB, f_x, f_y, f_w, f_h, screenshade);
+    V_DrawShaded(FB, f_x, f_y, f_w, f_h, SCREENSHADE);
 
   if (map_textured)
   {
diff --git a/prboom2/src/m_menu.c b/prboom2/src/m_menu.c
index ce58856f3..0cbccddf4 100644
--- a/prboom2/src/m_menu.c
+++ b/prboom2/src/m_menu.c
@@ -5835,8 +5835,6 @@ void M_StartControlPanel (void)
 //
 // This displays a dark overlay under certain screens of the menus
 
-int screenshade = 20;
-
 dboolean fadeBG(void)
 {
   return dsda_IntConfig(dsda_config_menu_background) == 1;
@@ -5850,7 +5848,7 @@ dboolean M_MenuIsShaded(void)
 
 static void M_ShadedScreen(int scrn)
 {
-  V_DrawShaded(scrn, 0, 0, SCREENWIDTH, SCREENHEIGHT, screenshade);
+  V_DrawShaded(scrn, 0, 0, SCREENWIDTH, SCREENHEIGHT, SCREENSHADE);
 }
 
 //
diff --git a/prboom2/src/m_menu.h b/prboom2/src/m_menu.h
index 2048b4000..a69bc267c 100644
--- a/prboom2/src/m_menu.h
+++ b/prboom2/src/m_menu.h
@@ -51,7 +51,7 @@ dboolean M_Responder (event_t *ev);
 
 dboolean fadeBG(void);
 dboolean M_MenuIsShaded(void);
-extern int screenshade;
+#define SCREENSHADE 20
 
 // Called by main loop,
 // only used for menu (skull cursor) animation.
```

Comment (andrikpowell) [reply prboom2/src/m_menu.c]: So in Nyan Doom, `screenshade` is a variable that changes due to some gradual fade logic. To keep this consistent between both DSDA and Nyan, I think I'll replace `int screenshade` here with a `#define FULLSHADE 20` which Nyan uses to compare `screenshade` to.

**Review by rfomin — APPROVED** [review]
LGTM, thank you.

**Comment by andrikpowell** [comment]
Ok, fixed up PR and simplified commits 👍

**Comment by Pedro-Beirao** [comment]
~~Please make Textured be the default mode for Menu Background. Like it was before~~

And I see that Textured only "kicks in" when going into sub-menus of Options, while Dark is straight when going into Options. Its inconsistent, but I kind of like it like this (and woof does the same). So no need to change this, just wanted to note it

**Review by Pedro-Beirao — APPROVED** [review]

**Comment by Pedro-Beirao** [comment]
Thanks!
